antenna/ext/pubsub/crashpublish.py

```diff
@@ -0,0 +1,84 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import logging
+
+from everett.manager import Option
+from google.cloud.pubsub_v1 import PublisherClient
+from google.cloud.pubsub_v1.types import BatchSettings
+
+from antenna.ext.crashpublish_base import CrashPublishBase
+from antenna.app import register_for_verification
+
+
+logger = logging.getLogger(__name__)
+
+
+class PubSubCrashPublish(CrashPublishBase):
+    """Publisher to Pub/Sub.
+
+    **Required GCP things**
+
+    To use this, you need to create:
+
+    1. Google Compute project
+    2. topic in that project
+    5. subscription for that topic so you can consume queued items
+
+    If something in the above isn't created, then Antenna may not start.
+
+
+    **Verification**
+
+    This component verifies that it can publish to the topic by publishing a
+    fake crash id of ``test``. Downstream consumer should throw this out.
+
+
+    **Local emulaior**
+
+    If you set the environment variable ``PUBSUB_EMULATOR_HOST=host:port``,
+    then this will connect to a local Pub/Sub emulator.
+
+    """
+
+    class Config:
+        project_id = Option(doc="Google Cloud project id.")
+        topic_name = Option(doc="The Pub/Sub topic name to publish to.")
+
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.project_id = self.config("project_id")
+        self.topic_name = self.config("topic_name")
+
+        # publish messages immediately without queuing.
+        batch_settings = BatchSettings(max_messages=1)
+        self.publisher = PublisherClient(batch_settings)
```

Comment:
How is this authenticating in GCP? Previously we were using a service account file.

Also, `PublisherClient` takes an optional `PublisherOptions` which lets us specify how many times to retry, timeouts, etc. The docs are kind of unhelpful and don't state what the defaults are for retry and timeout.

https://cloud.google.com/python/docs/reference/pubsub/latest/google.cloud.pubsub_v1.types.PublisherOptions

It also looks like it can take an `api_endpoint` argument. We could do what we did with sqs, make that a configuration option, and then pass that in in the case of the local dev environment. It's a little more explicit than setting an environment variable is.

Alternatively, if we want to stick with the environment variable, it'd be good for this to log something along the lines of "PUBSUB_EMULATOR_HOST is set--using emulator."

Reply:
> How is this authenticating in GCP?

per https://googleapis.dev/python/google-api-core/latest/auth.html:
> - If you’re running in a Google Virtual Machine Environment (Compute Engine, App Engine, Cloud Run, Cloud Functions), authentication should “just work”.
> - If you’re developing locally, the easiest way to authenticate is using the [Google Cloud SDK](http://cloud.google.com/sdk):
>   ```
>   $ gcloud auth application-default login
>   ```
> - If you’re running your application elsewhere, you should download a [service account](https://cloud.google.com/iam/docs/creating-managing-service-accounts#creating) JSON keyfile and point to it using an environment variable:
>   ```
>   $ export GOOGLE_APPLICATION_CREDENTIALS="/path/to/keyfile.json"
>   ```

I don't think we should maintain any additional code to explicitly add to those options.

> Also, PublisherClient takes an optional PublisherOptions which lets us specify how many times to retry, timeouts, etc. The docs are kind of unhelpful and don't state what the defaults are for retry and timeout.
> 
> https://cloud.google.com/python/docs/reference/pubsub/latest/google.cloud.pubsub_v1.types.PublisherOptions

the defaults appear in [google.cloud.pubsub_v1.types.PublishFlowControl](https://cloud.google.com/python/docs/reference/pubsub/latest/google.cloud.pubsub_v1.types.PublishFlowControl), message limit is 1,000 and byte limit is 10,000,000 but the limit exceeded behavior is IGNORE (effectively turns off flow control). The other options for limit exceeded behavior are block (wait until there is room) and error (reject messages). Note that these are not batch settings, but whole-client settings for limiting the amount of in-flight data. This should stick to the default of IGNORE.

> It also looks like it can take an api_endpoint argument. We could do what we did with sqs, make that a configuration option, and then pass that in in the case of the local dev environment. It's a little more explicit than setting an environment variable is.

as with auth, I don't think we should maintain anything in addition to the sdk's built-in support for specifying an emulator. Specifically for pubsub, in the past setting an alternative api endpoint was insufficient for connecting to the emulator, and in fact it was impossible to specify an emulator via options without the environment variable or manually recreating a bunch of code around instantiating grpc connections.

> Alternatively, if we want to stick with the environment variable, it'd be good for this to log something along the lines of "PUBSUB_EMULATOR_HOST is set--using emulator."

will do

Reply:
> > Also, PublisherClient takes an optional PublisherOptions which lets us specify how many times to retry, timeouts, etc. The docs are kind of unhelpful and don't state what the defaults are for retry and timeout.
> > https://cloud.google.com/python/docs/reference/pubsub/latest/google.cloud.pubsub_v1.types.PublisherOptions
> 
> the defaults appear in [google.cloud.pubsub_v1.types.PublishFlowControl](https://cloud.google.com/python/docs/reference/pubsub/latest/google.cloud.pubsub_v1.types.PublishFlowControl), message limit is 1,000 and byte limit is 10,000,000 but the limit exceeded behavior is IGNORE (effectively turns off flow control). The other options for limit exceeded behavior are block (wait until there is room) and error (reject messages). Note that these are not batch settings, but whole-client settings for limiting the amount of in-flight data. This should stick to the default of IGNORE.

looks like i got distracted an answered the wrong question. the only information i can find is "retry transient errors indefinitely": https://github.com/googleapis/python-pubsub/blob/efbaf617bafd6d5f04c0b0f183745547a1e95f44/google/cloud/pubsub_v1/publisher/client.py#L133-L135 but I think that doesn't account for the timeout. I'm not sure.

Reply:
transient errors are defined in google-api-core https://googleapis.dev/python/google-api-core/latest/_modules/google/api_core/retry.html

Reply:
Re: authentication, we need to document how authentication needs to be set up. This should be covered in two places:

1. in the Socorro GCP project plan because it's stuff we need to set up for the infrastructure
2. in the PubSubCrashPublish docstring in case someone else is using antenna--they need to know how to set it up

Re: retry and timeout, is it correct to say the way we've got it set up, it'll retry indefinitely and wait 120 seconds between retry attempts? Or is the timeout "times out after 120 seconds in connection or transmit" and not about retry?

Reply:
> Re: retry and timeout, is it correct to say the way we've got it set up, it'll retry indefinitely and wait 120 seconds between retry attempts? Or is the timeout "times out after 120 seconds in connection or transmit" and not about retry?

after rereading [the docs for the Retry class](https://googleapis.dev/python/google-api-core/latest/retry.html#google.api_core.retry.Retry) I see there are two timeouts that I was conflating.

120 seconds is the default retry timeout, and as implemented this means that no further attempts will be made after 120 seconds, but it does not interrupt an ongoing request.

the other is the rpc timeout, and is the one specified directly to [PublisherOptions](https://cloud.google.com/python/docs/reference/pubsub/latest/google.cloud.pubsub_v1.types.PublisherOptions). It is implemented in [google.api_core.timeout](https://googleapis.dev/python/google-api-core/latest/_modules/google/api_core/timeout.html), and defaults to None (does not timeout).

As I understand it: the default behavior is that the client may block indefinitely during connect and transmit, and will only timeout if it's actually getting responses from the pubsub server and those responses fall into the category of "transient errors".

Given the retry handling built in to antenna, I am going to change the default behavior to disable google's retries, and set an rpc timeout of 5 seconds. I'll include it in the docs with more details, but also mentioning here: in my past experience pubsub requests take less than 150ms, and if they take even 1s things are going horribly wrong (and the mozilla data platform's ingestion edge is at risk of failure, which has happened once since their gcp migration, mitigations were put in place, but those mitigations remain unproven because it hasn't happened again)